python/lsst/analysis/ap/plotImageSubtractionCutouts.py

```diff
@@ -50,6 +49,44 @@
 from . import apdb
 
 
+class ButlerCache:
+    """Global class to handle butler queries, to allow lru_cache and
+    multiprocessing.pool to work together.
+
+    If we redo this all to work with BPS or other parallelized systems, or get
+    good butler-side caching, we could remove this lru_cache system.
+    """
+    def set(self, butler, config):
+        """Call this to store a Butler and pex.Config instance before using the
+        global class instance.
+        """
+        self._butler = butler
+        self._config = config
+        # Ensure the caches are empty if we've been re-set.
+        self.get_exposures.cache_clear()
+        self.get_catalog.cache_clear()
+
+    @functools.lru_cache(maxsize=4)
+    def get_exposures(self, instrument, detector, visit):
+        """Return science, template, difference exposures, using a small
+        cache so we don't have to re-read files as often.
+        """
+        dataId = {'instrument': instrument, 'detector': detector, 'visit': visit}
+        return (self._butler.get(self._config.science_image_type, dataId),
+                self._butler.get(f'{self._config.diff_image_type}_templateExp', dataId),
+                self._butler.get(f'{self._config.diff_image_type}_differenceExp', dataId))
+
+    @functools.lru_cache(maxsize=4)
+    def get_catalog(self, instrument, detector, visit):
+        """Return the diaSrc catalog from the butler."""
+        dataId = {'instrument': instrument, 'detector': detector, 'visit': visit}
+        return self._butler.get(f'{self._config.diff_image_type}_diaSrc', dataId)
+
+
+# Global used within each multiprocessing worker (or single process).
+butler_cache = ButlerCache()
```

Comment:
I think this is basically working because the global environment is being copied on fork to a new sub-process, so all the symbols can be effectively seen. You may be able to solve this with an initializer in the pool jobs. I only bring this up because there was a case a few months ago about a global (thus import time) thing tripping us up somewhere else. I think it is probably fine here, but wanted to raise the concern.

Reply:
I couldn't get the initializer to work without the global here (for reasons that now escape me). I know I'm adding a global that gets copied on fork, and I'm not happy about it, but because we need to pass around Butler and Config instances, the initializer basically has to create a global variable anyway, as I understood it.